readme.txt
In respect to all those dying arcade games. Keep them alive, support the 
developers when you can. Sources should be included, have fun!

spice2x fork maintainers:
* sp2xdev

SpiceTools original contributors:
* cube
* felix
* dinsfire
* nolm
* s2
* crazyredmachine (jotaro44)
* mon
* nibs
* dogelition_man
* FSH
* itsnovahere
* arm1stice
* kinetic
* veroxzik
* AllanCat


changelog.txt
09/29/2023 [spice2x]
    Fix subscreen not updating in certain versions of EG.
    Detect long paths and log a warning message.

06/08/2023 [spice2x]
    Improved touch support for Jubeat and Reflec Beat.

04/23/2023 [spice2x]
    Update -iidxsounddevice to automatically pick a value (wasapi in most cases)

04/21/2023 [spice2x]
    Fix spicecfg not launching in some specific cases

04/20/2023 [spice2x]
    Update font for IIDX segment display
    New option: auto show FPS window

04/16/2023 [spice2x] (beta)
    New feature: window resize. Added new options to change window size and
        position on launch. Screen Resize window (F11) also updated with new
        controls
    Screen Resize settings are now per-game
    Misc bug fixes

04/09/2023 [spice2x] (beta)
    Add IIDX LED ticker (segment display) as subscreen overlay for LDJ
    Add I/O panel window for all games, special support for IIDX, DDR, GFDM
    Add options to automatically show certain windows on game launch

04/03/2023 [spice2x] (beta)
    Move/resize for IIDX/SDVX subscreen overlay window
    New options to control subscreen overlay window
    Turn common config mistakes into fatal error messages
    Various bug fixes for touch handling
    Revert "pop'n launching at desktop refresh rate" hack from 03/30/2023

03/30/2023 [spice2x]
    New option: IIDX Force Definition Type HD (720p)
    New option: process priority
    Fix pop'n launching at desktop refresh rate
    Lower logger thread priority to reduce stutters
    Misc bug fixes

03/25/2023 [spice2x]
    Screen Resize UI update (F11)
    FPS UI update (F12)
    New option: processor affinity
    Misc bug fixes

03/20/2023 [spice2x]
    ASIO device override for SDVX (experimental)
    Fix subscreen overlay for SDVX in Valkyrie mode using BI2X I/O
    Update text for touch and audio config options
    UI tweaks and minor bug fixes

03/17/2023 [spice2x]
    Disable UPX (file sizes will be larger)
    Fix regression from 2022 that prevented early patches
        from being loaded properly (such as audio hex edits)
    New option: HID lighting brightness
    New option: display auto-rotate direction
    New option: custom sleep duration

03/15/2023 [spice2x]
    Improve IIDX TT+/- handling (allow alternate key bind)
    Fix SpiceCompanion touch targets for TDJ FHD
    Config UI tweaks

03/13/2023 [spice2x]
    Fix TDJ FHD launching at wrong refresh rate
    New IIDX/SDVX options for TT/knob sensitivity when using digital input

03/12/2023 [spice2x]
    Fix mouse input for SDVX EG subscreen overlay
    New unified layout for Options tab
    Allow multiple DLLs in Inject Hook field

03/11/2023 [spice2x]
    Beginning of spice2x fork, based on another fork
    Ensure TDJ mode runs at 1080p@120Hz for newer games
    Fix IIDX subscreen overlay in FHD mode
    Gracefully deal with bad ea3-config.xml

04/02/2022 - Add lua scripting
             Add button edit dialog
             Add button debounce/invert settings
             Add bind many function
             Misc bug fixes
12/28/2021 - Add DirectX 9On12 wrapper option
             Add analog deadzone settings
             Fix SSL handshaking failed issue
             Misc bug fixes
12/19/2021 - Force single graphics adapter option
             Performance improvements
             Misc bug fixes
10/12/2021 - Ongaku Paradise support
             Fixed memory leak in Windows 8+ touch handler
             Fix Gitadora two channel fix for Nextage
             Fix Charge Machine touch window position
             IIDX 19 support
04/02/2021 - Charge Machine support
             Fix "unsupported fs type" endless loop
             Misc bug fixes
03/01/2021 - Add DDR HD speaker lights
             Automatic /dev/nvram and /dev/raw creation for AVS 2.17 and later
             Fix FlexASIO crash introduced in previous release
             Fix overlay crash from ImGui navigation button handling
             Fix AVS configuration reading with binary files
             Add AVS VFS drive remapping for E: and F:
             Improve XP compatibility
             Force dummy audio when using ASIO backend
             Misc bug fixes
12/30/2020 - Metal Gear Arcade lights and fixes
             Misc bug fixes
12/11/2020 - Fix patch manager crashing on invalid address
             Automatic /dev/nvram and /dev/raw creation for AVS 2.16 and prior
             Analog sensitivity improvements
             Locale improvements
             Misc bug fixes
10/27/2020 - Screen Capture support
             IIDX TDJ support
             Updated patches
             Misc bug fixes
10/13/2020 - Configurator patches tab
             KFControl fix cursor wrap
             Screen resize window
             Misc bug fixes
10/03/2020 - Add KFControl
             Misc bug fixes
09/27/2020 - Reflec limelight & colette (LBR) support
             LovePlus lights support
             Fix module loading when spice(64).exe is outside contents
             Ignore touchpad HID devices in touchscreen detection
             Keypad bindings
             SSL support
             Analog knob preview
06/15/2020 - QMA physical keyboard support
             Misc bug fixes
05/19/2020 - Alternate audio backends for exclusive-mode applications
             Fix screenshot handler crash in KMA and LMA
             MessageBox hooks (stops fullscreen windows from losing focus)
04/25/2020 - pop'n music 19 & 20 support
             DDR X2 support
             pop'n & HPM HDMI fix
             Misc additions, fixes
03/13/2020 - Misc bug fixes and improvements
03/10/2020 - Winning Eleven support
03/04/2020 - Tenkaichi Shogikai support
             Otoca D'or support
             Misc bug fixes
02/29/2020 - MIDI hotplugging / overlay
             Misc bug fixes
02/04/2020 - Configurator options tab
             Overlay button bindings
             Card generator
             Misc bug fixes
01/22/2020 - Audio improvements
             SDVX print overwrite option

misc/eamuse.cpp
#include "eamuse.h"

#include <fstream>
#include <thread>

#include "avs/game.h"
#include "cfg/config.h"
#include "games/io.h"
#include "rawinput/rawinput.h"
#include "games/sdvx/sdvx.h"
#include "util/logging.h"
#include "util/time.h"
#include "util/utils.h"

#include "bt5api.h"

// state
static bool CARD_INSERT[2] = {false, false};
static double CARD_INSERT_TIME[2] = {0, 0};
static double CARD_INSERT_TIMEOUT = 2.0;
static char CARD_INSERT_UID[2][8];
static char CARD_INSERT_UID_ENABLE[2] = {false, false};
static int COIN_STOCK = 0;
static bool COIN_BLOCK = false;
static std::thread *COIN_INPUT_THREAD;
static bool COIN_INPUT_THREAD_ACTIVE = false;
static uint16_t KEYPAD_STATE[] = {0, 0};
static uint16_t KEYPAD_STATE_OVERRIDES[] = {0, 0};
static uint16_t KEYPAD_STATE_OVERRIDES_BT5[] = {0, 0};
static uint16_t KEYPAD_STATE_OVERRIDES_READER[] = {0, 0};
static uint16_t KEYPAD_STATE_OVERRIDES_OVERLAY[] = {0, 0};
static std::string EAMUSE_GAME_NAME;
static ConfigKeypadBindings KEYPAD_BINDINGS {};

bool eamuse_get_card(int active_count, int unit_id, uint8_t *card) {

    // get unit index
    int index = unit_id > 0 && active_count > 1 ? 1 : 0;

    // reader card input
    if (CARD_INSERT_UID_ENABLE[index]) {
        CARD_INSERT_UID_ENABLE[index] = false;
        memcpy(card, CARD_INSERT_UID[index], 8);
        log_info("eamuse", "Inserted card from reader {}: {}", index, bin2hex(card, 8));
        return true;
    }

    // get file path
    std::filesystem::path path;
    if (!KEYPAD_BINDINGS.card_paths[index].empty()) {
        path = KEYPAD_BINDINGS.card_paths[index];
    } else {
        path = index > 0 ? "card1.txt" : "card0.txt";
    }

    // call the next function
    return eamuse_get_card(path, card, index);
}

bool eamuse_get_card(const std::filesystem::path &path, uint8_t *card, int index) {

    // Check if card overrides are present
    if (!CARD_OVERRIDES[index].empty()) {

        // Override is present
        for (int n = 0; n < 16; n++) {
            char c = CARD_OVERRIDES[index].c_str()[n];
            bool digit = c >= '0' && c <= '9';
            bool character_big = c >= 'A' && c <= 'F';
            bool character_small = c >= 'a' && c <= 'f';
            if (!digit && !character_big && !character_small) {
                log_warning("eamuse",
                            "{} card override contains an invalid character sequence at byte {} (16 characters, 0-9/A-F only)",
                            CARD_OVERRIDES[index], n);

                return false;
            }
        }
        // Log info
        log_info("eamuse", "Inserted card override: {}", CARD_OVERRIDES[index]);

        // Card is valid, convert and set it.
        hex2bin(CARD_OVERRIDES[index].c_str(), card);

        // success
        return true;
    }
    // Overrides are not present. Use the standard file reading method.
    return eamuse_get_card(path, card);
}

bool eamuse_get_card(const std::filesystem::path &path, uint8_t *card) {

    // open file
    std::ifstream f(path);
    if (!f) {
        log_warning("eamuse", "{} can not be opened!", path.string());
        return false;
    }

    // get size
    f.seekg(0, f.end);
    auto length = (size_t) f.tellg();
    f.seekg(0, f.beg);

    // check size
    if (length < 16) {
        log_warning("eamuse", "{} is too small (must be at least 16 characters)", path.string());
        return false;
    }

    // read file
    char buffer[17];
    f.read(buffer, 16);
    buffer[16] = 0;

    // verify card
    for (int n = 0; n < 16; n++) {
        char c = buffer[n];
        bool digit = c >= '0' && c <= '9';
        bool character_big = c >= 'A' && c <= 'F';
        bool character_small = c >= 'a' && c <= 'f';
        if (!digit && !character_big && !character_small) {
            log_warning("eamuse",
                "{} contains an invalid character sequence at byte {} (16 characters, 0-9/A-F only)",
                path.string(), n);

            return false;
        }
    }

    // info
    log_info("eamuse", "Inserted {}: {}", path.string(), buffer);

    // convert hex to bytes
    hex2bin(buffer, card);

    // success
    return true;
}

void eamuse_card_insert(int unit) {
    CARD_INSERT[unit] = true;
    CARD_INSERT_TIME[unit] = get_performance_seconds();
}

void eamuse_card_insert(int unit, const uint8_t *card) {
    memcpy(CARD_INSERT_UID[unit], card, 8);
    CARD_INSERT[unit] = true;
    CARD_INSERT_TIME[unit] = get_performance_seconds();
    CARD_INSERT_UID_ENABLE[unit] = true;
}

bool eamuse_card_insert_consume(int active_count, int unit_id) {

    // get unit index
    int index = unit_id > 0 && active_count > 1 ? 1 : 0;

    // bt5api
    if (BT5API_ENABLED) {
        bt5api_poll_reader_card((uint8_t) index);
    }

    // check for card insert
    auto keypad_buttons = games::get_buttons_keypads(eamuse_get_game());
    auto offset = unit_id * games::KeypadButtons::Size;
    if ((CARD_INSERT[index] && fabs(get_performance_seconds() - CARD_INSERT_TIME[index]) < CARD_INSERT_TIMEOUT)
    || GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::InsertCard + offset))) {
        log_info("eamuse", "Card insert on {}/{}", unit_id + 1, active_count);
        CARD_INSERT[index] = false;
        return true;
    }

    return false;
}

bool eamuse_coin_get_block() {
    return COIN_BLOCK;
}

void eamuse_coin_set_block(bool block) {
    COIN_BLOCK = block;
}

int eamuse_coin_get_stock() {
    return COIN_STOCK;
}

void eamuse_coin_set_stock(int amount) {
    COIN_STOCK = amount;
}

bool eamuse_coin_consume(int amount) {
    if (COIN_STOCK < amount) {
        return false;
    } else {
        COIN_STOCK -= amount;
        return true;
    }
}

int eamuse_coin_consume_stock() {
    int stock = COIN_STOCK;
    COIN_STOCK = 0;
    return stock;
}

int eamuse_coin_add() {
    return ++COIN_STOCK;
}

void eamuse_coin_start_thread() {

    // set active
    COIN_INPUT_THREAD_ACTIVE = true;

    // create thread
    COIN_INPUT_THREAD = new std::thread([]() {
        auto overlay_buttons = games::get_buttons_overlay(eamuse_get_game());
        static bool COIN_INPUT_KEY_STATE = false;
        while (COIN_INPUT_THREAD_ACTIVE) {

            // check input key
            if (overlay_buttons && GameAPI::Buttons::getState(RI_MGR, overlay_buttons->at(
                    games::OverlayButtons::InsertCoin))) {
                if (!COIN_INPUT_KEY_STATE) {
                    if (COIN_BLOCK)
                        log_info("eamuse", "coin inserted while blocked");
                    else {
                        log_info("eamuse", "coin insert");
                        COIN_STOCK++;
                    }
                }
                COIN_INPUT_KEY_STATE = true;
            } else {
                COIN_INPUT_KEY_STATE = false;
            }

            // once every two frames
            Sleep(1000 / 30);
        }
    });
}

void eamuse_coin_stop_thread() {
    COIN_INPUT_THREAD_ACTIVE = false;
    COIN_INPUT_THREAD->join();
    delete COIN_INPUT_THREAD;
    COIN_INPUT_THREAD = nullptr;
}

void eamuse_set_keypad_overrides(size_t unit, uint16_t keypad_state) {

    // check unit
    if (unit >= std::size(KEYPAD_STATE_OVERRIDES)) {
        return;
    }

    // set state
    KEYPAD_STATE_OVERRIDES[unit] = keypad_state;
}

void eamuse_set_keypad_overrides_bt5(size_t unit, uint16_t keypad_state) {

    // check unit
    if (unit >= std::size(KEYPAD_STATE_OVERRIDES_BT5)) {
        return;
    }

    // set state
    KEYPAD_STATE_OVERRIDES_BT5[unit] = keypad_state;
}

void eamuse_set_keypad_overrides_reader(size_t unit, uint16_t keypad_state) {

    // check unit
    if (unit >= std::size(KEYPAD_STATE_OVERRIDES_READER)) {
        return;
    }

    // set state
    KEYPAD_STATE_OVERRIDES_READER[unit] = keypad_state;
}

void eamuse_set_keypad_overrides_overlay(size_t unit, uint16_t keypad_state) {

    // check unit
    if (unit >= std::size(KEYPAD_STATE_OVERRIDES_OVERLAY)) {
        return;
    }

    // set state
    KEYPAD_STATE_OVERRIDES_OVERLAY[unit] = keypad_state;
}

uint16_t eamuse_get_keypad_state(size_t unit) {

    // check unit
    if (unit >= std::size(KEYPAD_STATE)) {
        return 0;
    }
    
    // keypad disabled for KFC in BI2X mode, since the game won't handle it correctly
    // this check for IO is necessary (as opposed to just doing a check for Valkyrie cab mode)
    // because there are hex edits that allow you to use legacy (KFC/BIO2) IO while in Valk mode.
    if (avs::game::is_model("KFC") && games::sdvx::BI2X_INITIALIZED) {
        return 0;
    }

    // reset
    KEYPAD_STATE[unit] = KEYPAD_STATE_OVERRIDES[unit];
    KEYPAD_STATE[unit] |= KEYPAD_STATE_OVERRIDES_BT5[unit];
    KEYPAD_STATE[unit] |= KEYPAD_STATE_OVERRIDES_READER[unit];
    KEYPAD_STATE[unit] |= KEYPAD_STATE_OVERRIDES_OVERLAY[unit];

    // bt5api
    if (BT5API_ENABLED) {
        bt5api_poll_reader_keypad((uint8_t) unit);
    }

    // get keybinds and get offset for unit
    auto keypad_buttons = games::get_buttons_keypads(eamuse_get_game());
    auto offset = unit * games::KeypadButtons::Size;

    // parse
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad0 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_0;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad1 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_1;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad2 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_2;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad3 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_3;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad4 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_4;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad5 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_5;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad6 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_6;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad7 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_7;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad8 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_8;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad9 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_9;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::Keypad00 + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_00;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::KeypadDecimal + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_KEYPAD_DECIMAL;
    }
    if (GameAPI::Buttons::getState(RI_MGR, keypad_buttons->at(games::KeypadButtons::InsertCard + offset))) {
        KEYPAD_STATE[unit] |= 1 << EAM_IO_INSERT;
    }

    // return state
    return KEYPAD_STATE[unit];
}

std::string eamuse_get_keypad_state_str(size_t unit) {
    auto state = eamuse_get_keypad_state(unit);
    std::ostringstream ss;
    if (state & 1 << EAM_IO_KEYPAD_0) ss << "0";
    if (state & 1 << EAM_IO_KEYPAD_1) ss << "1";
    if (state & 1 << EAM_IO_KEYPAD_2) ss << "2";
    if (state & 1 << EAM_IO_KEYPAD_3) ss << "3";
    if (state & 1 << EAM_IO_KEYPAD_4) ss << "4";
    if (state & 1 << EAM_IO_KEYPAD_5) ss << "5";
    if (state & 1 << EAM_IO_KEYPAD_6) ss << "6";
    if (state & 1 << EAM_IO_KEYPAD_7) ss << "7";
    if (state & 1 << EAM_IO_KEYPAD_8) ss << "8";
    if (state & 1 << EAM_IO_KEYPAD_9) ss << "9";
    if (state & 1 << EAM_IO_KEYPAD_00) ss << "00";
    if (state & 1 << EAM_IO_KEYPAD_DECIMAL) ss << "00";
    if (state & 1 << EAM_IO_INSERT) ss << "00";
    return ss.str();
}

bool eamuse_keypad_state_naive() {
    return KEYPAD_BINDINGS.keypads[0].empty() && KEYPAD_BINDINGS.keypads[1].empty();
}

void eamuse_set_game(std::string game) {
    if (EAMUSE_GAME_NAME != game) {
        EAMUSE_GAME_NAME = std::move(game);
        eamuse_update_keypad_bindings();
    }
}

void eamuse_update_keypad_bindings() {
    KEYPAD_BINDINGS = Config::getInstance().getKeypadBindings(EAMUSE_GAME_NAME);
}

std::string eamuse_get_game() {
    return EAMUSE_GAME_NAME;
}

int eamuse_get_game_keypads() {
    if (avs::game::is_model("JDZ") ||
        avs::game::is_model("KDZ") ||
        avs::game::is_model("LDJ") ||
        avs::game::is_model("JDX") ||
        avs::game::is_model("KDX") ||
        avs::game::is_model("MDX") ||
        avs::game::is_model("J33") ||
        avs::game::is_model("K33") ||
        avs::game::is_model("L33") ||
        avs::game::is_model("M32"))
    {
        return 2;
    }

    return 1;
}

int eamuse_get_game_keypads_name() {
    auto game_name = eamuse_get_game();

    if (game_name == "Beatmania IIDX" ||
        game_name == "Dance Dance Revolution" ||
        game_name == "GitaDora")
    {
        return 2;
    }

    return 1;
}

void eamuse_autodetect_game() {
    if (avs::game::is_model("KFC"))
        eamuse_set_game("Sound Voltex");
    else if (avs::game::is_model("JDZ") ||
             avs::game::is_model("KDZ") ||
             avs::game::is_model("LDJ"))
        eamuse_set_game("Beatmania IIDX");
    else if (avs::game::is_model("J44") ||
             avs::game::is_model("K44") ||
             avs::game::is_model("L44"))
        eamuse_set_game("Jubeat");
    else if (avs::game::is_model("KDM"))
        eamuse_set_game("Dance Evolution");
    else if (avs::game::is_model("NBT"))
        eamuse_set_game("Beatstream");
    else if (avs::game::is_model("I36"))
        eamuse_set_game("Metal Gear");
    else if (avs::game::is_model("KBR") ||
             avs::game::is_model("LBR") ||
             avs::game::is_model("MBR"))
        eamuse_set_game("Reflec Beat");
    else if (avs::game::is_model("KBI"))
        eamuse_set_game("Tenkaichi Shogikai");
    else if (avs::game::is_model("K39") ||
             avs::game::is_model("L39") ||
             avs::game::is_model("M39"))
        eamuse_set_game("Pop'n Music");
    else if (avs::game::is_model("KGG"))
        eamuse_set_game("Steel Chronicle");
    else if (avs::game::is_model("JGT"))
        eamuse_set_game("Road Fighters 3D");
    else if (avs::game::is_model("PIX"))
        eamuse_set_game("Museca");
    else if (avs::game::is_model("R66"))
        eamuse_set_game("Bishi Bashi Channel");
    else if (avs::game::is_model("J32") ||
             avs::game::is_model("J33") ||
             avs::game::is_model("K32") ||
             avs::game::is_model("K33") ||
             avs::game::is_model("L32") ||
             avs::game::is_model("L33") ||
             avs::game::is_model("M32"))
        eamuse_set_game("GitaDora");
    else if (avs::game::is_model("JDX") ||
             avs::game::is_model("KDX") ||
             avs::game::is_model("MDX"))
        eamuse_set_game("Dance Dance Revolution");
    else if (avs::game::is_model("PAN"))
        eamuse_set_game("Nostalgia");
    else if (avs::game::is_model("JMA") ||
             avs::game::is_model("KMA") ||
             avs::game::is_model("LMA"))
        eamuse_set_game("Quiz Magic Academy");
    else if (avs::game::is_model("MMD"))
        eamuse_set_game("FutureTomTom");
    else if (avs::game::is_model("KK9"))
        eamuse_set_game("Mahjong Fight Club");
    else if (avs::game::is_model("JMP"))
        eamuse_set_game("HELLO! Pop'n Music");
    else if (avs::game::is_model("KLP"))
        eamuse_set_game("LovePlus");
    else if (avs::game::is_model("NSC"))
        eamuse_set_game("Scotto");
    else if (avs::game::is_model("REC"))
        eamuse_set_game("DANCERUSH");
    else if (avs::game::is_model("KCK") ||
             avs::game::is_model("NCK"))
        eamuse_set_game("Winning Eleven");
    else if (avs::game::is_model("NCG"))
        eamuse_set_game("Otoca D'or");
    else if (avs::game::is_model("LA9"))
        eamuse_set_game("Charge Machine");
    else if (avs::game::is_model("JC9"))
        eamuse_set_game("Ongaku Paradise");
    else if (avs::game::is_model("TBS"))
        eamuse_set_game("Busou Shinki: Armored Princess Battle Conductor");
    else {
        log_warning("eamuse", "unknown game model: {}", avs::game::MODEL);
        eamuse_set_game("unknown");
    }
}

void eamuse_scard_callback(uint8_t slot_no, card_info_t *cardinfo) {
    if (cardinfo->card_type > 0) {
        eamuse_card_insert(slot_no, cardinfo->uid);
    }
}
